### PR TITLE
Prepend snowflake to title upon tab discard

### PR DESCRIPTION
**Pull request by Rob--W**
This PR restores the feature that you intended to have: a snowflake in the tab title.

Previously, you tried to update the title via `tabs.update`. That is not possible, because the tabs.update API does not support a "title" key (#5), so it was removed (#6).

This PR does the following:

- Add logic to prepend a snowflake to the tab title before discarding it.
- Add options page to enable the snowflake (with an optional permission prompt)
- Explicitly add the extension ID to manifest.json (found by visitting about:support after installing the add-on from AMO), so that the extension storage is persisted during development.

An intentional side effect of the implementation is that a snowflake briefly appears and then disappears if you attempt to discard the currently selected tab. The current tab cannot be discarded by design, so this piece of UI feedback shows that something has happened in response to the user's request to discard the tab.

### CODE REVIEW

**Comment by Rob--W** [comment]
PS.  the optional permission should not have been necessary, if the `activeTab` permission worked as expected. This is however not the case: https://bugzilla.mozilla.org/show_bug.cgi?id=1446956
If that Firefox bug is fixed, the extension can be updated to use the activeTab permission instead of `<all_urls>`, and then the runtime permission prompts are not necessary any more.

**Comment by freddyb** [comment]
Hey Rob,
thanks for this pull request. The code is really clean!

That being said, I'm a bit hesitant to add the `"<all_urls>"` permission for this tiny, cosmetic feature. I will give this some more thought and testing on a later day.

**Comment by Rob--W** [comment]
This PR uses optional permissions, so there is no install-time warning.
Users who don't want to use the feature don't have to enable it, while
users who do want the feature can grant the permission via the options page.

(And if the Firefox bug from my second comment is fixed, then there is no
need for any extra permission warnings).

**Comment by Rob--W** [comment]
@freddyb Have you already evaluated the PR? It is currently a bit difficult to know whether a tab has really been discarded.

**Comment by Rob--W** [comment]
@freddyb Ping?

File: manifest.json [hunk 1]
```diff
@@ -11,13 +11,23 @@
   },
   "applications": {
     "gecko": {
+      "id": "{09420bd6-3c35-444d-b85c-74d51fe3df4f}",
```
Comment (freddyb) [inline manifest.json]: Why is this necessary?

File: options.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<head>
+<meta charset="utf-8">
+</head>
+<body>
+<label>
+  <input type="checkbox" id="prefPrependSnowflake">
+  <!-- options.js will add a label with i18n ID toggleSnowflakePref. -->
```
Comment (freddyb) [inline options.html]: Can we put the text in a `<span>` that defaults to hardcoded English and gets changed to the localized version via script? Replacing `textContent` via the script might save us a layout reflow (and give a nicer experience to the default locale)

File: options.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+"use strict";
+
+var permissions = { origins: ["<all_urls>"] };
+var prefPrependSnowflake = document.getElementById("prefPrependSnowflake");
+prefPrependSnowflake.onchange = async () => {
+  if (prefPrependSnowflake.checked) {
+    prefPrependSnowflake.disabled = true;
+    // Note: permissions.request only works when this page is shown in a tab,
+    // i.e. when options_ui.open_in_tab=true in manifest.json.
+    // See https://bugzilla.mozilla.org/show_bug.cgi?id=1382953
```
Comment (freddyb) [inline options.js:10]: Once this has landed, I'd like us to track this in an issue for the repo here.

**Review by freddyb — COMMENTED** [review]
First of all, thank you for this pull request. As I mentioned before, I recommend we talk about some course of action in an issue before you go ahead.

I really like how this panned out (and apologize for the huge delay in reviewing this!).

I have a few questions before I want to approve this.

File: manifest.json (hunk 1 above)
Comment (Rob--W) [reply manifest.json]: I added this in because `browser.storage.sync` requires an extension ID. However, I don't use `browser.storage.sync`, only `browser.storage.local`, so not this can be removed if you wish.

Do you want to switch to the `storage.sync` API or continue to use `storage.local` ?

File: options.js (hunk 3 above)
Comment (Rob--W) [reply options.js:10]: This upstream bug is going to land soon, possibly fixed in Firefox 61.

File: manifest.json (hunk 1 above)
Comment (freddyb) [reply manifest.json]: I'd prefer `storage.local`. What do you think?

File: options.html (hunk 2 above)
Comment (Rob--W) [reply options.html]: Done. Initially I did not do so to not duplicate the string, but since you've requested it, I have amended the commit. This is the difference:

```diff
diff --git a/options.html b/options.html
index 9eea240..a9c2d09 100644
--- a/options.html
+++ b/options.html
@@ -5,7 +5,8 @@
 <body>
 <label>
   <input type="checkbox" id="prefPrependSnowflake">
-  <!-- options.js will add a label with i18n ID toggleSnowflakePref. -->
+  <!-- options.js will replace the label with i18n ID toggleSnowflakePref. -->
+  <span id="i18n-toggleSnowflakePref">Prepend snowflake to discarded tabs (requires extra permission).</span>
 </label>
 <script src="options.js"></script>
 </body>
diff --git a/options.js b/options.js
index 173c3f0..57cdc40 100644
--- a/options.js
+++ b/options.js
@@ -18,8 +18,8 @@ prefPrependSnowflake.onchange = async () => {
   });
 };
 
-prefPrependSnowflake.insertAdjacentText("afterEnd",
-  browser.i18n.getMessage("toggleSnowflakePref"));
+document.getElementById("i18n-toggleSnowflakePref").textContent =
+  browser.i18n.getMessage("toggleSnowflakePref");
 
 browser.storage.onChanged.addListener((changes) => {
   if (changes.prependSnowflake) {
```

File: manifest.json (hunk 1 above)
Comment (Rob--W) [reply manifest.json]: SGTM. I have amended the commit to remove this ID.

**Comment by freddyb** [comment]
:sparkles:

**Comment by freddyb** [comment]
sorry this has been moving so slowly. I'll probably make a release sometime next week.
